libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_RangeFinder_config.h"
+
+#if AP_RANGEFINDER_JRE_SERIAL_ENABLED
+
+#include "AP_RangeFinder_JRE_Serial.h"
+#include <AP_Math/AP_Math.h>
+
+#define FRAME_HEADER_1   'R'    // 0x52
+#define FRAME_HEADER_2   'A'    // 0x41
+
+#define DIST_MAX_CM 50000
+#define OUT_OF_RANGE_ADD_CM   100
+
+bool AP_RangeFinder_JRE_Serial::get_reading(float &reading_m)
+{
+    // uart instance check
+    if (uart == nullptr) {
+        return false;  // not update
+    }
+
+    uint16_t valid_count = 0;   // number of valid readings
+    uint16_t invalid_count = 0; // number of invalid readings
+    float sum = 0;
+    // max distance the sensor can reliably measure - read from parameters
+    const int16_t distance_cm_max = max_distance_cm();
+
+    // buffer read
+    const ssize_t num_read = uart->read(read_buff, ARRAY_SIZE(read_buff));
+    read_buff_idx = 0; // read_buff start data index
+    while (read_buff_idx < num_read) {
```

Comment:
I agree with Randy's comments.

If READ_BUFF_IDX is incremented in the condition of the WHILE statement, the loop will not be infinite.
If there is an increment leak during processing within the loop, the loop will go around indefinitely.

In other range finders, the index is updated under the condition of the WHILE statement.

Reply:
@muramura 

I changed the 'while loop' to a 'for loop'.

and
I added the codes to adapt additional protocols of this sensor.
See header file (AP_RangeFinder_JRE_Serial.h).